File: Sources/FitUI/Components/FitVideoUploadCard.swift
import SwiftUI

// MARK: - FitVideoUploadCard
//
// 16:9 state-aware video upload card. Owns the visual lifecycle of a
// direct-upload video integration (Mux is the primary consumer):
//   .idle       — dashed dropzone with camera icon + CTA
//   .uploading  — yellow-tinted card with linear progress + cancel ×
//   .processing — yellow-tinted card with spinner
//   .ready      — solid card with thumbnail + play overlay + ⋯ menu
//   .errored    — red-tinted card with alert + retry CTA
//   .pending    — same visual as .idle + "last upload didn't finish" hint
//
// Owners pass platform integration (picker, upload, player) via callbacks.
// The kit component does NOT pick files, perform the PUT, or play video
// itself — that's app-level work that varies per platform.
//
// See `docs/components.md` § FitVideoUploadCard.

public struct FitVideoUploadCard: View {
    public enum State: Equatable {
        case idle
        case uploading(progress: Double, filename: String)
        case processing
        case ready(thumbnailURL: URL?)
        case errored(errorCode: String?)
        case pending
    }

    let state: State
    let onTap: () -> Void
    let onCancel: (() -> Void)?
    let onMore: (() -> Void)?

    @Environment(\.fitTheme) private var theme

    public init(
        state: State,
        onTap: @escaping () -> Void,
        onCancel: (() -> Void)? = nil,
        onMore: (() -> Void)? = nil
    ) {
        self.state = state
        self.onTap = onTap
        self.onCancel = onCancel
        self.onMore = onMore
    }

    public var body: some View {
        ZStack(alignment: .topTrailing) {
            content
                .frame(maxWidth: .infinity)
                .aspectRatio(16.0 / 9.0, contentMode: .fit)
                .background(backgroundFill)
                .overlay(borderOverlay)
                .clipShape(RoundedRectangle(cornerRadius: FitRadius.md))
                .contentShape(Rectangle())
                .onTapGesture(perform: handleTap)

            overlayControls
                .padding(.top, FitSpacing.sp2)
                .padding(.trailing, FitSpacing.sp2)
        }
    }

    // MARK: - Body content by state

    @ViewBuilder
    private var content: some View {
        switch state {
        case .idle, .pending:
            idleSlot
        case .uploading(let progress, let filename):
            uploadingSlot(progress: progress, filename: filename)
        case .processing:
            processingSlot
        case .ready(let thumbnailURL):
            readySlot(thumbnailURL: thumbnailURL)
        case .errored(let errorCode):
            erroredSlot(errorCode: errorCode)
        }
    }

    // MARK: - IDLE / PENDING

    private var idleSlot: some View {
        VStack(spacing: FitSpacing.sp2) {
            Image(systemName: "video")
                .font(.system(size: 24, weight: .regular))
                .foregroundStyle(theme.textSecondary.opacity(0.6))
                .frame(width: 32, height: 32)
            Text("Upload intro video")
                .font(.custom(FitFont.family, size: 15).weight(.medium))
                .foregroundStyle(theme.textPrimary)
            Text("Up to 200 MB · 2 min · mp4/mov")
                .font(.custom(FitFont.family, size: 12))
                .foregroundStyle(theme.textTertiary)
        }
        .multilineTextAlignment(.center)
        .padding(FitSpacing.sp3)
    }

    // MARK: - UPLOADING

    private func uploadingSlot(progress: Double, filename: String) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 6) {
                Image(systemName: "video.fill")
                    .font(.system(size: 12))
                Text(filename)
                    .font(.custom(FitFont.family, size: 13))
                    .lineLimit(1)
                    .truncationMode(.middle)
            }
            .foregroundStyle(theme.textPrimary)

            // Progress bar
            GeometryReader { proxy in
                ZStack(alignment: .leading) {
                    Capsule()
                        .fill(theme.surfaceHigh)
                    Capsule()
                        .fill(FitColors.warning)
                        .frame(width: max(0, min(1, progress)) * proxy.size.width)
                }
            }
            .frame(height: 6)

            Text("Uploading… \(Int(max(0, min(1, progress)) * 100))%")
                .font(.custom(FitFont.family, size: 12))
                .foregroundStyle(theme.textTertiary)
                .frame(maxWidth: .infinity, alignment: .trailing)
        }
        .padding(.horizontal, FitSpacing.sp4)
        .padding(.vertical, 14)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    // MARK: - PROCESSING

    private var processingSlot: some View {
        VStack(spacing: 10) {
            ProgressView()
                .progressViewStyle(.circular)
                .tint(FitColors.warning)
                .scaleEffect(1.2)
                .frame(width: 28, height: 28)
            Text("Processing your video…")
                .font(.custom(FitFont.family, size: 14).weight(.medium))
                .foregroundStyle(FitColors.warning)
            Text("Usually 30-60 seconds")
                .font(.custom(FitFont.family, size: 12))
                .foregroundStyle(theme.textTertiary)
        }
        .multilineTextAlignment(.center)
        .padding(FitSpacing.sp3)
    }

    // MARK: - READY

    private func readySlot(thumbnailURL: URL?) -> some View {
        ZStack {
            if let thumbnailURL {
                AsyncImage(url: thumbnailURL) { phase in
                    switch phase {
                    case .success(let image):
                        image.resizable().scaledToFill()
                    default:
                        FitColors.brandGradient
                    }
                }
            } else {
                FitColors.brandGradient
            }

            // Play overlay
            Color.black.opacity(0.25)
            Image(systemName: "play.fill")
                .font(.system(size: 36))
                .foregroundStyle(.white)
                .shadow(color: .black.opacity(0.4), radius: 6, y: 2)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    // MARK: - ERRORED

    private func erroredSlot(errorCode: String?) -> some View {
        VStack(spacing: 6) {
            Image(systemName: "exclamationmark.triangle")
                .font(.system(size: 24, weight: .regular))
                .foregroundStyle(FitColors.error)
                .frame(width: 32, height: 32)
            Text("Couldn’t process this video")
                .font(.custom(FitFont.family, size: 14).weight(.medium))
                .foregroundStyle(theme.textPrimary)
            Text("Try a different file")
                .font(.custom(FitFont.family, size: 14).weight(.medium))
                .foregroundStyle(FitColors.error)
                .underline()
            if let errorCode {
                Text("code: \(errorCode)")
                    .font(.custom(FitFont.family, size: 11))
                    .foregroundStyle(theme.textTertiary)
                    .padding(.top, 2)
            }
        }
        .multilineTextAlignment(.center)
        .padding(FitSpacing.sp3)
    }

    // MARK: - Top-right controls (cancel × in uploading, ⋯ in ready)

    @ViewBuilder
    private var overlayControls: some View {
        switch state {
        case .uploading:
            if let onCancel {
                Button(action: onCancel) {
                    Image(systemName: "xmark")
                        .font(.system(size: 12, weight: .semibold))
                        .foregroundStyle(.white)
                        .frame(width: 28, height: 28)
                        .background(Color.black.opacity(0.6), in: Circle())
                }
                .buttonStyle(.plain)
            }
        case .ready:
            if let onMore {
                Button(action: onMore) {
                    Image(systemName: "ellipsis")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(.white)
                        .frame(width: 32, height: 32)
                        .background(Color.black.opacity(0.5), in: Circle())
                }
                .buttonStyle(.plain)
            }
        default:
            EmptyView()
        }
    }

    // MARK: - State-driven styling

    private var backgroundFill: Color {
        switch state {
        case .idle, .pending:
            return .clear
        case .uploading, .processing:
            return theme.bgWarningSubtle
        case .ready:
            return theme.surfaceHigher
        case .errored:
            return theme.bgErrorSubtle
        }
    }

    @ViewBuilder
    private var borderOverlay: some View {
        let shape = RoundedRectangle(cornerRadius: FitRadius.md)
        switch state {
        case .idle, .pending:
            shape.strokeBorder(theme.divider, style: StrokeStyle(lineWidth: 1.5, dash: [5, 4]))
        case .uploading, .processing:
            shape.strokeBorder(FitColors.warning, lineWidth: 1)
        case .ready:
            shape.strokeBorder(theme.divider, lineWidth: 1)
        case .errored:
            shape.strokeBorder(FitColors.error, lineWidth: 1)
        }
    }

    // MARK: - Tap routing

    private func handleTap() {
        switch state {
        case .idle, .pending, .errored, .ready:
            onTap()
        case .uploading, .processing:
            // Card is non-interactive; cancel × handles abort during upload.
            break
        }
    }
}

// MARK: - Preview

#if DEBUG
struct FitVideoUploadCard_Previews: PreviewProvider {
    static var previews: some View {
        ScrollView {
            VStack(spacing: 16) {
                FitVideoUploadCard(state: .idle, onTap: {})
                FitVideoUploadCard(state: .uploading(progress: 0.64, filename: "intro-2026.mp4"), onTap: {}, onCancel: {})
                FitVideoUploadCard(state: .processing, onTap: {})
                FitVideoUploadCard(state: .ready(thumbnailURL: nil), onTap: {}, onMore: {})
                FitVideoUploadCard(state: .errored(errorCode: "invalid_input"), onTap: {})
                FitVideoUploadCard(state: .pending, onTap: {})
            }
            .padding()
        }
        .background(Color.black)
        .fitTheme(.coach)
        .previewDisplayName("Coach (dark) — all states")
    }
}
#endif

File: android/src/main/kotlin/com/fit321/fitui/components/FitVideoUploadCard.kt
package com.fit321.fitui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.MoreHoriz
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Videocam
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.res.vectorResource
import com.fit321.designtokens.R
import com.fit321.fitui.theme.LocalFitTheme
import com.fit321.fitui.tokens.FitColors
import com.fit321.fitui.tokens.FitFont

// ============================================================================
// FitVideoUploadCard — 16:9 state-aware video upload card for direct-upload
// integrations (Mux is the primary consumer). Mirrors Swift `FitVideoUploadCard`.
// See `docs/components.md` § FitVideoUploadCard.
// ============================================================================

sealed class FitVideoUploadCardState {
    object Idle : FitVideoUploadCardState()
    data class Uploading(val progress: Float, val filename: String) : FitVideoUploadCardState()
    object Processing : FitVideoUploadCardState()
    data class Ready(val thumbnailUrl: String?) : FitVideoUploadCardState()
    data class Errored(val errorCode: String?) : FitVideoUploadCardState()
    object Pending : FitVideoUploadCardState()
}

@Composable
fun FitVideoUploadCard(
    state: FitVideoUploadCardState,
    onTap: () -> Unit,
    onCancel: (() -> Unit)? = null,
    onMore: (() -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    val theme = LocalFitTheme.current
    val shape = RoundedCornerShape(12.dp)

    val isTappable = when (state) {
        is FitVideoUploadCardState.Uploading,
        is FitVideoUploadCardState.Processing -> false
        else -> true
    }

    val tapModifier = if (isTappable) Modifier.clickable { onTap() } else Modifier

    // Compose Modifier.border doesn't natively support dashed strokes for
    // RoundedCornerShape; iOS uses StrokeStyle dash. Android uses solid
    // border at v1; switch to drawWithCache custom paint if dashed becomes
    // a strict requirement later (e.g. accessibility audit).
    val borderColor: Color
    val borderWidth: androidx.compose.ui.unit.Dp
    val backgroundColor: Color

    when (state) {
        is FitVideoUploadCardState.Idle,
        is FitVideoUploadCardState.Pending -> {
            borderColor = theme.divider
            borderWidth = 1.5.dp
            backgroundColor = Color.Transparent
        }
        is FitVideoUploadCardState.Uploading,
        is FitVideoUploadCardState.Processing -> {
            borderColor = FitColors.warning
            borderWidth = 1.dp
            backgroundColor = theme.bgWarningSubtle
        }
        is FitVideoUploadCardState.Ready -> {
            borderColor = theme.divider
            borderWidth = 1.dp
            backgroundColor = theme.surfaceHigher
        }
        is FitVideoUploadCardState.Errored -> {
            borderColor = FitColors.error
            borderWidth = 1.dp
            backgroundColor = theme.bgErrorSubtle
        }
    }

    Box(
        modifier = modifier
            .fillMaxWidth()
            .aspectRatio(16f / 9f)
            .clip(shape)
            .background(backgroundColor)
            .border(borderWidth, borderColor, shape)
            .then(tapModifier)
    ) {
        when (state) {
            is FitVideoUploadCardState.Idle,
            is FitVideoUploadCardState.Pending -> IdleSlot(theme)
            is FitVideoUploadCardState.Uploading -> UploadingSlot(state.progress, state.filename, theme)
            is FitVideoUploadCardState.Processing -> ProcessingSlot(theme)
            is FitVideoUploadCardState.Ready -> ReadySlot()
            is FitVideoUploadCardState.Errored -> ErroredSlot(state.errorCode, theme)
        }

        // Top-right overlay controls
        when (state) {
            is FitVideoUploadCardState.Uploading -> if (onCancel != null) {
                CornerButton(Alignment.TopEnd, size = 28.dp, onClick = onCancel) {
                    Icon(ImageVector.vectorResource(R.drawable.ic_fit_close), null, tint = Color.White, modifier = Modifier.size(14.dp))
                }
            }
            is FitVideoUploadCardState.Ready -> if (onMore != null) {
                CornerButton(Alignment.TopEnd, size = 32.dp, onClick = onMore) {
                    Icon(Icons.Default.MoreHoriz, null, tint = Color.White, modifier = Modifier.size(18.dp))
                }
            }
            else -> { /* no overlay */ }
        }
    }
}

@Composable
private fun BoxScope.IdleSlot(theme: FitColors.Theme) {
    Column(
        modifier = Modifier.align(Alignment.Center).padding(12.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Icon(
            Icons.Default.Videocam,
            contentDescription = null,
            tint = theme.textSecondary.copy(alpha = 0.6f),
            modifier = Modifier.size(32.dp)
        )
        Text(
            "Upload intro video",
            style = FitFont.body1.copy(fontWeight = FontWeight.Medium, fontSize = 15.sp),
            color = theme.textPrimary,
            textAlign = TextAlign.Center
        )
        Text(
            "Up to 200 MB · 2 min · mp4/mov",
            style = FitFont.caption.copy(fontSize = 12.sp),
            color = theme.textTertiary,
            textAlign = TextAlign.Center
        )
    }
}

@Composable
private fun BoxScope.UploadingSlot(
    progress: Float,
    filename: String,
    theme: FitColors.Theme
) {
    val clamped = progress.coerceIn(0f, 1f)
    Column(
        modifier = Modifier
            .align(Alignment.Center)
            .fillMaxSize()
            .padding(horizontal = 16.dp, vertical = 14.dp),
        verticalArrangement = Arrangement.Center
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp)
        ) {
            Icon(
                Icons.Default.Videocam,
                contentDescription = null,
                tint = theme.textPrimary,
                modifier = Modifier.size(14.dp)
            )
            Text(
                filename,
                style = FitFont.body2.copy(fontSize = 13.sp),
                color = theme.textPrimary,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
        }
        Spacer(Modifier.height(10.dp))
        // Progress bar — 6.dp track with yellow fill scaled by progress
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(6.dp)
                .clip(RoundedCornerShape(3.dp))
                .background(theme.surfaceHigh)
        ) {
            Box(
                modifier = Modifier
                    .fillMaxWidth(clamped)
                    .fillMaxSize()
                    .background(FitColors.warning)
            )
        }
        Spacer(Modifier.height(6.dp))
        Text(
            "Uploading… ${(clamped * 100).toInt()}%",
            style = FitFont.caption.copy(fontSize = 12.sp),
            color = theme.textTertiary,
            textAlign = TextAlign.End,
            modifier = Modifier.fillMaxWidth()
        )
    }
}

@Composable
private fun BoxScope.ProcessingSlot(theme: FitColors.Theme) {
    Column(
        modifier = Modifier.align(Alignment.Center).padding(12.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp)
    ) {
        CircularProgressIndicator(
            color = FitColors.warning,
            strokeWidth = 2.5.dp,
            modifier = Modifier.size(28.dp)
        )
        Text(
            "Processing your video…",
            style = FitFont.body1.copy(fontWeight = FontWeight.Medium, fontSize = 14.sp),
            color = FitColors.warning,
            textAlign = TextAlign.Center
        )
        Text(
            "Usually 30-60 seconds",
            style = FitFont.caption.copy(fontSize = 12.sp),
            color = theme.textTertiary,
            textAlign = TextAlign.Center
        )
    }
}

@Composable
private fun BoxScope.ReadySlot() {
    // Real impl: ExoPlayer or AsyncImage from Coil with thumbnailUrl.
    // Prototype-level placeholder = brand-gradient fill + centered play overlay.
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(FitColors.brandGradient)
    ) {
        // Dim scrim
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.Black.copy(alpha = 0.25f))
        )
        Icon(
            Icons.Default.PlayArrow,
            contentDescription = null,
            tint = Color.White,
            modifier = Modifier
                .align(Alignment.Center)
                .size(48.dp)
        )
    }
}

@Composable
private fun BoxScope.ErroredSlot(
    errorCode: String?,
    theme: FitColors.Theme
) {
    Column(
        modifier = Modifier.align(Alignment.Center).padding(12.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(6.dp)
    ) {
        Icon(
            Icons.Default.Warning,
            contentDescription = null,
            tint = FitColors.error,
            modifier = Modifier.size(32.dp)
        )
        Text(
            "Couldn’t process this video",
            style = FitFont.body1.copy(fontWeight = FontWeight.Medium, fontSize = 14.sp),
            color = theme.textPrimary,
            textAlign = TextAlign.Center
        )
        Text(
            "Try a different file",
            style = FitFont.body1.copy(fontWeight = FontWeight.Medium, fontSize = 14.sp),
            color = FitColors.error,
            textAlign = TextAlign.Center
        )
        if (errorCode != null) {
            Text(
                "code: $errorCode",
                style = FitFont.caption.copy(fontSize = 11.sp),
                color = theme.textTertiary,
                textAlign = TextAlign.Center
            )
        }
    }
}

@Composable
private fun BoxScope.CornerButton(
    alignment: Alignment,
    size: androidx.compose.ui.unit.Dp,
    onClick: () -> Unit,
    content: @Composable () -> Unit
) {
    Box(
        modifier = Modifier
            .align(alignment)
            .padding(8.dp)
            .size(size)
            .clip(CircleShape)
            .background(Color.Black.copy(alpha = 0.55f))
            .clickable { onClick() },
        contentAlignment = Alignment.Center
    ) {
        content()
    }
}
